core/operation/src/main/java/uk/gov/gchq/gaffer/operation/impl/Repeat.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2017 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.gov.gchq.gaffer.operation.impl;
+
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import com.fasterxml.jackson.core.type.TypeReference;
+import org.apache.commons.lang3.exception.CloneFailedException;
+
+import uk.gov.gchq.gaffer.operation.Operation;
+import uk.gov.gchq.gaffer.operation.Operations;
+import uk.gov.gchq.gaffer.operation.io.InputOutput;
+import uk.gov.gchq.gaffer.operation.serialisation.TypeReferenceImpl;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@code Repeat} is a Gaffer {@link Operation} for executing a delegate <code>Operation</code>
+ * for a specified number of repeats.
+ * <p>The delegate operation can also be an implementation of {@link Operations},
+ * so for example, a small {@link uk.gov.gchq.gaffer.operation.OperationChain} could be delegated
+ * to the <code>Repeat</code>, and repeated a number of times.</p>
+ */
+@JsonPropertyOrder(value = {"class", "input", "operation", "times"}, alphabetic = true)
+public class Repeat implements InputOutput<Object, Object>,
```

Comment:
Does this only make sense if the operation being repeated also implements ```InputOutput<I, O>```? If so, the class signature could be amended to:

`java
public class Repeat<I, O> implements InputOutput<I, O>, Operations<InputOutput<I,O>
`

This then would indicate that the Input/Output types are the same:

`java
public class Repeat<T> implements InputOutput<T, T>, Operations<InputOutput<T, T>
`

Perhaps there is room for two of these operations, one which is more general and one which provides a bit more type safety? Or does that introduce unnecessary complication?

Reply:
I kind of agree with: public class Repeat<T> implements InputOutput<T, T>. However, if the operation is designed to repeat, then I think the input type must be the same as the output type. However It is possible to repeat a GetAdjacentIds operation. The input and output types are actually slightly different. Similarly with GetElements. So we probably will need to use public class Repeat<I, O> implements InputOutput<I, O>

Regarding this: Operations<InputOutput<I,O>   we do this repeat to work with any type of operation. For example repeatedly doing AddElements (no output). Or repeatedly running GetAllElements (no input). Or it could be a chain of operations, consisting of some operations with an input, some with an output and some with both. So I think we will have to leave it as Operations<Operation>.